### PR TITLE
Add dask lazy loading

### PR DESCRIPTION
**Pull request by michalk8**
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (change that would cause existing functionality to not work as before)
- [ ] New feature (non-breaking change which adds functionality)

### Description

Add lazy loading for JPG/PNG using dask.array. Also allows loading for Z-dim, currently disabled.

### How has this been tested?

~~No tests.~~
Modify 1 test to test that we can now load PNGs.

### Closes

...

TODO:
- [x] finish error messages

### CODE REVIEW

File: squidpy/im/_io.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    fname = str(fname)
+    shape, c_dim, z_dim, dtype = _get_shape_pages(fname)
+    dims, shape = _determine_dimensions(shape, c_dim, z_dim)
+
+    darr = da.from_delayed(delayed(imread)(fname).reshape(shape), shape=shape, dtype=dtype)
+    if chunks is not None:
+        darr = darr.rechunk(chunks)
+
+    # TODO: subsetting for bwd compat
+    return xr.DataArray(darr, dims=dims).transpose("y", "x", "z", "channels")[:, :, 0, :]
```
Comment (michalk8) [inline squidpy/im/_io.py:92]: @hspitzer you can uncomment the `[...]` so that, might be a starting point for #321

Comment (hspitzer) [reply squidpy/im/_io.py:99]: awesome, thanks for setting that up already!

File: squidpy/im/_io.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+def _lazy_load_image(fname: Union[str, Path], chunks: Optional[str] = None) -> xr.DataArray:
+    from dask import delayed
+    import dask.array as da
+
+    fname = str(fname)
+    shape, c_dim, z_dim, dtype = _get_shape_pages(fname)
+    dims, shape = _determine_dimensions(shape, c_dim, z_dim)
+
+    darr = da.from_delayed(delayed(imread)(fname).reshape(shape), shape=shape, dtype=dtype)
```
Comment (hspitzer) [inline squidpy/im/_io.py:94]: This is a really nice and elegant way of reading images. Great! I have to admit that I did not fully understand the `_get_shape_pages` and `_determine_dimensions` functions, but the code works on all examples that I tried it on. Like magic 👍

**Comment by giovp** [comment]
hey, ust wanted to try it out on new data we got from collaborators and it's still an issue.
If I try to load jpg
```python
img = sq.im.ImageContainer(
    "./../../dat/SP01-2020_GE_HE_JSON/GE1_V19S18-084_CA1A.jpg",
)
```

error is

<details>
<summary>Details</summary>

```
/Users/giovanni.palla/miniconda3/envs/spatial/lib/python3.8/site-packages/pims/api.py:207: UserWarning: <class 'pims.image_reader.ImageReader'> errored: Image size (1029959493 pixels) exceeds limit of 178956970 pixels, could be decompression bomb DOS attack.
  warn(message)
/Users/giovanni.palla/miniconda3/envs/spatial/lib/python3.8/site-packages/pims/api.py:207: UserWarning: <class 'pims.image_reader.ImageReaderND'> errored: Image size (1029959493 pixels) exceeds limit of 178956970 pixels, could be decompression bomb DOS attack.
  warn(message)
/Users/giovanni.palla/miniconda3/envs/spatial/lib/python3.8/site-packages/pims/api.py:207: UserWarning: <class 'pims.imageio_reader.ImageIOReader'> errored: The ImageIOReader requires imageio and (for imageio >= 2.5) imageio-ffmpeg to work.
  warn(message)
/Users/giovanni.palla/miniconda3/envs/spatial/lib/python3.8/site-packages/pims/api.py:207: UserWarning: <class 'pims.bioformats.BioformatsReader'> errored: 'NoneType' object has no attribute 'isJVMStarted'
  warn(message)
---------------------------------------------------------------------------
UnknownFormatError                        Traceback (most recent call last)
<ipython-input-2-4e20ba4bb5d2> in <module>
----> 1 img = sq.im.ImageContainer(
      2     "./../../dat/SP01-2020_GE_HE_JSON/GE1_V19S18-084_CA1A.jpg",
      3 )

~/Projects/squidpy/squidpy/im/_container.py in __init__(self, img, layer, **kwargs)
    101             if chunks is not None:
    102                 chunks = {"x": chunks, "y": chunks}
--> 103             self.add_img(img, layer=layer, chunks=chunks, **kwargs)
    104 
    105     @classmethod

~/Projects/squidpy/squidpy/im/_container.py in add_img(self, img, layer, channel_dim, lazy, chunks, **kwargs)
    197         """
    198         layer = self._get_next_image_id("image") if layer is None else layer
--> 199         img = self._load_img(img, chunks=chunks, layer=layer, **kwargs)
    200 
    201         if img is not None:  # not reading a .nc file

~/miniconda3/envs/spatial/lib/python3.8/functools.py in _method(*args, **kwargs)
    910         def _method(*args, **kwargs):
    911             method = self.dispatcher.dispatch(args[0].__class__)
--> 912             return method.__get__(obj, cls)(*args, **kwargs)
    913 
    914         _method.__isabstractmethod__ = self.__isabstractmethod__

~/Projects/squidpy/squidpy/im/_container.py in _(self, img, chunks, **_)
    246 
    247         if suffix in (".jpg", ".jpeg", ".png", ".tif", ".tiff"):
--> 248             return _lazy_load_image(img, chunks=None)
    249 
    250         if img.is_dir():

~/Projects/squidpy/squidpy/im/_io.py in _lazy_load_image(fname, chunks)
     89 
     90     fname = str(fname)
---> 91     shape, c_dim, z_dim, dtype = _get_shape_pages(fname)
     92     dims, shape = _determine_dimensions(shape, c_dim, z_dim)
     93 

~/Projects/squidpy/squidpy/im/_io.py in _get_shape_pages(fname)
     13     import pims
     14 
---> 15     with pims.open(fname) as img:
     16         shape = [len(img)] + list(img.frame_shape)
     17         dtype = np.dtype(img.pixel_type)

~/miniconda3/envs/spatial/lib/python3.8/site-packages/pims/api.py in open(sequence, **kwargs)
    207             warn(message)
    208             exceptions += message + '\n'
--> 209     raise UnknownFormatError("All handlers returned exceptions:\n" + exceptions)
    210 
    211 

UnknownFormatError: All handlers returned exceptions:
<class 'pims.image_reader.ImageReader'> errored: Image size (1029959493 pixels) exceeds limit of 178956970 pixels, could be decompression bomb DOS attack.
<class 'pims.image_reader.ImageReaderND'> errored: Image size (1029959493 pixels) exceeds limit of 178956970 pixels, could be decompression bomb DOS attack.
<class 'pims.imageio_reader.ImageIOReader'> errored: The ImageIOReader requires imageio and (for imageio >= 2.5) imageio-ffmpeg to work.
<class 'pims.bioformats.BioformatsReader'> errored: 'NoneType' object has no attribute 'isJVMStarted'
```

</details>

If I set the limit: `PIL.Image.MAX_IMAGE_PIXELS = 1029959493` then it loads but the kernel freezes (exceeding memory).

if I convert it to tiff and then load it as that, it works but the kernel freezes when I launch interactive:
```python
img = sq.im.ImageContainer(
    "./../../dat/SP01-2020_GE_HE_JSON/GE1_V19S18-084_CA1A.tiff",
)
adata = sc.read("./../../res/GE1_V19S18-084_CA1Aa.h5ad")
import napari
with napari.gui_qt():
    img.interactive(adata)
```

it's quite a large image
image

shall I send you the image to test it out @michalk8 ?

**Comment by michalk8** [comment]
@giovp , yes, that would be great.

**Comment by giovp** [comment]
> @giovp , yes, that would be great. Or I can just generate random RGB image.

coming very soon

**Comment by michalk8** [comment]
I've tried this:
```python
img = sq.im.ImageContainer(
    "./../../dat/SP01-2020_GE_HE_JSON/GE1_V19S18-084_CA1A.tiff",
)
adata = sc.read("./../../res/GE1_V19S18-084_CA1Aa.h5ad")
import napari
with napari.gui_qt():
    v = napari.Viewer()
    v.add_image(img['image'])
    # img.interactive(adata)
```
works fine, with a warning that GL_TEXTURE is only ~ 2^14 and I've had 30k x 30k image.

Furthermore, can confirm that when using `img.interactive(adata)`, there's huge mem. spike (~9GiB to 32+), because of `img_as_float`, which is always called when new layer is added.

**Comment by michalk8** [comment]
Also for large images, I will look into https://napari.org/tutorials/fundamentals/image.html#image-pyramids
and I'd say we should enable it for some large images (maybe larger that 16k x 16k.

**Comment by giovp** [comment]
ok interesting, but why this works
```
import napari
with napari.gui_qt():
    v = napari.Viewer()
    v.add_image(img['image'])
```
and not `interactive` ?

**Comment by michalk8** [comment]
> and not interactive ?

Because there's `img_as_float` call in `img.interactive`. I've removed it for now and enabled as well the pyramids for image with more than 65536 ** 2 elements.

**Comment by giovp** [comment]
> Because there's img_as_float call in img.interactive. I've removed it for now and enabled as well the pyramids for image with more than 65536 ** 2 elements.

amazing, can confirm I can load and open napari, both with tif and jpeg (although with jpeg need to set the PIL flag above).
I'll try to work on it a bit and report here,

thanks. alot for the prompt fixes @michalk8 !

**Comment by giovp** [comment]
ok, managed to segment tissues with napari, still impressed by how much functionality you put in the napari interface @michalk8 ...

**Comment by michalk8** [comment]
@giovp I've added the labels layer (e.g. watershed segmentation) + more control over how the image is displayed, i.e.
1 channel img. is displayed as grayscale (not sure about this one) and treating the channels as luminance, not RGB, see the code below.
The above changes should work with Z-stack, just need to uncomment stuff.
Will merge this soon to #324 and continue from there with watershed+dask.

```python
import napari
import squidpy as sq

adata = sq.datasets.visium_fluo_adata_crop()
img = sq.datasets.visium_fluo_image_crop()
_ = img.data.load()  # TODO: there's some padding problem with dask

sq.im.process(img, method="smooth")
sq.im.process(img, method="gray", layer='image_smooth')

img['image'].attrs['luminance'] = True  # treat as luminance, not as rg
# img['segmentation_watershed'].attrs['segmentation'] = True  # only works for integer-based images

with napari.gui_qt():
    img.interactive(adata)
```

**Comment by michalk8** [comment]
merged into #324

**Pull request closed without merging**